File: lab02/lab02/Problem.cs
namespace lab02;

public class Problem
{
    public List<Item> Items { get; set; }

    public Problem(List<Item> items)
    {
        Items = items;
    }
    
    public Problem(int n, int seed)
    {
        if (n < 0)
            throw new ArgumentException("Number of items cannot be negative");
        
        Items = new List<Item>();
        var random = new Random(seed);
        
        for (var i = 0; i < n; i++)
        {
            var value = random.Next(1, 10);
            var weight = random.Next(1, 10);
            Items.Add(new Item(i, value, weight));
        }
    }

    public Result Solve(int capacity)
    {
        if (capacity < 0)
        {
            throw new ArgumentException("Capacity cannot be negative");
        }
        
        if (Items.Count == 0)
            return new Result([]);
        
        var sorted = Items.OrderByDescending(item => item.Ratio()).ToList();
        var packedItems = new List<Item>();
        foreach (var itm in sorted)
            if (itm.Weight <= capacity)
            {
                packedItems.Add(itm);
                capacity -= itm.Weight;
            }
        return new Result(packedItems);
    }
    
    public override string ToString()
    {
        var itemsString = string.Join(", ", Items);
        return $"Problem(n={Items.Count}, items=[{itemsString}])";
    }
}

File: lab06/Knapsack/src/main/java/com/raczu/knapsack/Problem.java
package com.raczu.knapsack;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Random;

public class Problem {
    public List<Item> items;

    public Problem(List<Item> items) {
        this.items = items;
    }

    public Problem(int n, int seed, int lowerBound, int upperBound) throws IllegalArgumentException {
        if (n < 0) {
            throw new IllegalArgumentException("Number of items cannot be negative");
        }

        this.items = new ArrayList<>();
        Random random = new Random(seed);

        for (int i = 0; i < n; i++) {
            int value = random.nextInt(lowerBound, upperBound) + 1;
            int weight = random.nextInt(lowerBound, upperBound) + 1;
            this.items.add(new Item(i, value, weight));
        }
    }

    public Result solve(int capacity) throws IllegalArgumentException {
        if (capacity < 0) {
            throw new IllegalArgumentException("Capacity cannot be negative");
        }

        if (this.items.isEmpty()) {
            return new Result(List.of());
        }

        List<Item> sorted = new ArrayList<>(this.items);
        sorted.sort(Comparator.comparing(Item::ratio).reversed());
        List<Item> packed = new ArrayList<>();

        for (Item item : sorted) {
            while (item.weight <= capacity) {
                packed.add(item);
                capacity -= item.weight;
            }
        }

        return new Result(packed);
    }

    @Override
    public String toString() {
        return String.join("\n", this.items.stream().map(Item::toString).toList());
    }
}
